### ISSUE TITLE
[🐞] Nested onGet calls only return data from one onGet

### ISSUE THREAD

**Issue by tuurbo** [BuilderIO/qwik#2324]
### Which component is affected?

Qwik City (routing)

### Describe the bug

I'm in the process of moving a large, completely custom ecommerce store from Angular 14 to Qwik while reusing our graphql api. My biggest struggle right now is accessing cookies and setting document head meta tags which can only be done from onGet.

Nesting multiple onGet's only returns data from the deepest onGet call. For example in my ecommerce store the base /layout.tsx file uses onGet to get the customers cart item count and auth status. The /product/index.tsx page uses onGet to fetch the product details and set the document head like title, meta tags, etc. useResource$ doesn't work for me because i need access to cookies like you get from onGet.

### Reproduction

https://stackblitz.com/edit/qwik-starter-3zbwqa?file=src%2Froutes%2Flayout.tsx,src%2Froutes%2Findex.tsx

### System Info

```shell
System:
    OS: Windows 10 10.0.19045
    CPU: (20) x64 12th Gen Intel(R) Core(TM) i9-12900HK
    Memory: 16.29 GB / 31.68 GB
  Binaries:
    Node: 18.12.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.19 - C:\Program Files\nodejs\yarn.CMD
    npm: 8.19.2 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (107.0.1418.62)
    Internet Explorer: 11.0.19041.1566
  npmPackages:
    @builder.io/partytown: ^0.7.2 => 0.7.2
    @builder.io/qwik: 0.14.1 => 0.14.1
    @builder.io/qwik-city: 0.0.127 => 0.0.127
    vite: 3.2.4 => 3.2.4
```

**Comment by tuurbo** [comment]
The problem could possible be solved If `useResource$` had access to `RequestEvent` like onGet does. Then i wouldnt need to call onGet from the layout.tsx file.
```ts
const resource = useResource$(async ({ track, cookie, request, response }) => {
  // fetch cart and user
  return {
    cart: '...',
    user: '...',
  }
});
  ```



Or if multiple onGet's are called, then the data from each onGet is merged

```ts
// layout.tsx
export const onGet: RequestHandler = async (ctx) => {
  return {
    auth: '...',
    cart: '...'
  }
}

//  layout.tsx component$
useEndpoint<typeof onGet>();
{
  auth: '...',
  cart: '...',
}
```
```ts
// product/index.tsx
export const onGet: RequestHandler = async (ctx) => {
  return {
    product: '...'
  }
}

// product/index.tsx component$
useEndpoint<typeof onGet>();
{
  auth: '...',
  cart: '...',
  product: '...',
}
```

**Comment by adamdbradley** [comment]
This is something @manucorporat and I have been discussing, and because of the issues you're experiencing, we're leaning toward having layouts only act as middleware. Meaning it can get/set headers, change response types and abort going through each layout and last endpoint, however, a layout should be restricted from returning data. Think this is something we need to spec out a little bit more and would love keep the discussion going here on various ideas. Thanks

**Comment by jwickers** [comment]
A workaround for passing (limited?) data would be (ab)using the request headers.

For example in your layout onGet:
```ts
    const layoutData = {
        foo: 'bar ' + new Date().toISOString(),
    }
    request.headers.set('x-layout', JSON.stringify(layoutData))
```
then in a page onGet:
```ts
    const layoutHeader = request.headers.get('x-layout')
    const layoutData = layoutHeader ? JSON.parse(layoutHeader) : null
```

**Comment by KenAKAFrosty** [comment]
I for one prefer @jwickers idea of still offering this functionality during the chain of onGet handlers rather than blocking it off (if possible!).  This is a pretty common thing to run into and is often desired by devs. Would be a nice win for Qwik to have an easy way to lean into it (again, if possible and there's not a fundamental reason why it needs to be avoided!).

Recognizing that layouts effectively "wrap" all the routes at its level and deeper ends up being a natural place to want to do request handling like that. The term "layout" feels a _little_ odd for request handling, but other than that, the architecture itself as actually really really really really nice.  If possible, it'd be really nice to preserve that as a place to put logic like that. Because I mean where else would you?

Similar to @jwickers idea, something like `requestEvent.context` is by default an empty POJO. But you can add things to it in layout get handlers and it's passed down the chain. Then any given page onGet handler can read from it as needeed

**Comment by Nefcanto** [comment]
I wanted to create another issue with this [StackBlitz sample](https://stackblitz.com/edit/qwik-starter-zq7tfc?file=src%2Froutes%2Flayout.tsx), but when I searched the `onGet` I found this issue.

We also encourage you to not cut the data fetching possibility for layouts.
 
Because layouts usually load their own data. Like menu items, footer links, branding, certificates, etc.

And each page that is inside that layout, loads its own data. For example, the home page can load hero slides, services, pricing tables, testimonials, team members, latest blog posts, etc. while the about us page would load google Maps for location, contact form, and phone numbers and emails, and addresses.

If you disable data fetching for layouts, then we end up in the hell of Next.js 12. There, we had to load everything for each page and then use weird workarounds to pass data to parent layouts (_app.js in that case).

This means that we will be forced to write separate APIs per page (per URL) in our backend for our websites and we also have to load the whole layout data for each URL navigation.

Next.js 13 realized this shortcoming and they now let developers basically load data from layouts, as well as components.

That is very very flexible. So I strongly encourage you to not disable data fetching for layouts.

Thank you so much.
